swarm/src/behaviour/peer_addresses.rs

```diff
@@ -0,0 +1,227 @@
+use crate::behaviour::FromSwarm;
+use crate::{DialError, DialFailure, NewExternalAddrOfPeer};
+
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+
+use lru::LruCache;
+
+use std::collections::HashSet;
+use std::num::NonZeroUsize;
+
+/// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
+#[derive(Debug)]
+pub struct PeerAddresses(LruCache<PeerId, HashSet<Multiaddr>>);
+
+impl PeerAddresses {
+    /// Feed a [`FromSwarm`] event to this struct.
+    ///
+    /// Returns whether the event changed peer's known external addresses.
+    pub fn on_swarm_event(&mut self, event: &FromSwarm) -> bool {
+        match event {
+            FromSwarm::NewExternalAddrOfPeer(NewExternalAddrOfPeer { peer_id, addr }) => {
+                if let Some(peer_addrs) = self.get_mut(peer_id) {
+                    let addr = Self::prepare_addr(peer_id, addr);
+                    return peer_addrs.insert(addr);
+                } else {
+                    let addr = Self::prepare_addr(peer_id, addr);
+                    self.put(*peer_id, std::iter::once(addr));
+                    return true;
+                }
+            }
+            FromSwarm::DialFailure(DialFailure { peer_id, error, .. }) => {
+                if let Some(peer_id) = peer_id {
+                    match error {
+                        DialError::NoAddresses => return self.0.pop(peer_id).is_some(),
+                        _ => return false,
+                    }
+                }
+            }
```

Comment:
Collapsing patterns reduces indentation which IMO is easier to read:

```suggestion
            FromSwarm::DialFailure(DialFailure { peer_id: Some(peer_id), error: DialError::NoAddresses, .. }) => {
                return self.0.pop(peer_id).is_some()
            }
```

Reply:
:heavy_check_mark: